### PR TITLE
feat!: send diagnostics to Log Analytics

### LINKED ISSUE #61: feat(database)!: send diagnostics to Log Analytics workspace

### PR DESCRIPTION
**Pull request by kov117**
BREAKING CHANGE: add variable `log_analytics_workspace_id`.

### CODE REVIEW

File: main.tf [hunk 1]
```diff
@@ -19,46 +19,6 @@ module "server" {
...
-moved {
-  from = azurerm_storage_container.this
-  to   = module.server.azurerm_storage_container.this
-}
-
-moved {
-  from = azurerm_mssql_server_vulnerability_assessment.this
-  to   = module.server.azurerm_mssql_server_vulnerability_assessment.this
-}
-
```
Comment (hknutsen) [inline main.tf]: Don't remove these in public modules, that would make it very complicated to upgrade from previous versions as it would require manually running the `terraform state mv` command for each removed block.

File: variables.tf [hunk 2]
```diff
@@ -140,6 +140,36 @@ variable "long_term_retention_policy_week_of_year" {
...
+    "QueryStoreRuntimeStatistics",
+    "QueryStoreWaitStatistics",
+    "Errors",
+    "DatabaseWaitStatistics",
+    "Timeouts",
+    "Blocks",
+    "Deadlocks",
+    "DevOpsOperationsAudit",
+    "SQLSecurityAuditEvents"
+  ]
```
Comment (hknutsen) [inline variables.tf]: Set only to the log categories that are required for audit logging:

```suggestion
  default = [
    "DevOpsOperationsAudit",
    "SQLSecurityAuditEvents"
  ]
```

File: modules/database/variables.tf [hunk 3]
```diff
@@ -62,6 +62,42 @@ variable "long_term_retention_policy_week_of_year" {
...
+    "QueryStoreRuntimeStatistics",
+    "QueryStoreWaitStatistics",
+    "Errors",
+    "DatabaseWaitStatistics",
+    "Timeouts",
+    "Blocks",
+    "Deadlocks",
+    "DevOpsOperationsAudit",
+    "SQLSecurityAuditEvents"
+  ]
```
Comment (hknutsen) [inline modules/database/variables.tf]: Set only the log categories required for audit logging:

```suggestion
  default = [
    "DevOpsOperationsAudit",
    "SQLSecurityAuditEvents"
  ]
```

**Review by hknutsen — DISMISSED** [review]
General comments:

- I know the original issue #61 describes this as a feature for the database submodule, however it was concluded in that same issue that the audit logs that should be sent to Log Analytics are the same audit logs that are already being sent to Storage. The policy for sending audit logs to Storage was configured at the server-level, thus the policy for sending audit logs to Log Analytics should also be configured at the server-level.

  This can be done by configuring the diagnostic setting for the automatically created master database. See the Bicep example linked in the original issue. You can use a data source in Terraform to retrieve information about this automatically created master database.

- Which log categories to enable are usually optional, however in this case we always want the audit logs. Thus, these should be hardcoded. Consider adding additional log categories as a variable instead. (or, figure out a smart way to enable/disable audit logs 😉 )

- Finally, I'd rename the PR to be similar to the original issue: `feat!: send diagnostics to Log Analytics workspace`. Notice how i mark it as a breaking change. Try to figure out why, and include the changes needed to migrate in the PR description 😉

File: modules/server/main.tf [hunk 4]
```diff
@@ -58,11 +58,9 @@ resource "azurerm_mssql_firewall_rule" "this" {
 }
 
 resource "azurerm_mssql_server_extended_auditing_policy" "this" {
-  server_id                               = azurerm_mssql_server.this.id
-  storage_endpoint                        = azurerm_storage_account.this.primary_blob_endpoint
-  storage_account_access_key              = azurerm_storage_account.this.primary_access_key
-  storage_account_access_key_is_secondary = false
-  retention_in_days                       = 7
+  server_id              = azurerm_mssql_server.this.id
+  log_monitoring_enabled = true
```
Comment (helenakallekleiv) [inline modules/server/main.tf]:
```suggestion
  log_monitoring_enabled = true
  storage_endpoint                        = azurerm_storage_account.this.primary_blob_endpoint
  storage_account_access_key              = azurerm_storage_account.this.primary_access_key
  storage_account_access_key_is_secondary = false
```

@hknutsen @kov117 What are your thoughs on enabling log monitoring in addition to sending the extended server auditing policy to storage account? Or is that concidered as an unneccesary duplication?

Comment (hknutsen) [reply modules/server/main.tf]: Unnecessary duplication imo. 🙂

File: main.tf (hunk 1 above)
Comment (kov117) [reply main.tf]: ok, the moved blocks are back in place for future dokumentation. 👍🏽

File: modules/server/main.tf (hunk 4 above)
Comment (kov117) [reply modules/server/main.tf]: Resolving this conversation without suggested changes.

**Comment by kov117** [comment]
All of the issues have been resolved and fixed, however a new issue has emerged and i am not able to resolve it. The issue has to do with database master, the terratest cannot find the resource that is the "master" database, i have tried to resolve the issue with the use of lifecycle precondition, and using a different scripting method but the test result always ends up as error. The "master" database is to slow to ble created so whenever the test is running it cannot find the "master" database and therfore stops the test and declares it cannot find the resource.



#### Issue

```pwsh
Error: creating Monitor Diagnostics Setting "audit-logs" for Resource "/subscriptions/**/resourceGroups/**/providers/Microsoft.Sql/servers/**/databases/master": diagnosticsettings.DiagnosticSettingsClient#CreateOrUpdate: Failure responding to request: StatusCode=404 -- Original Error: autorest/azure: Service returned an error. Status=404 Code="ResourceNotFound" Message="The Resource 'Microsoft.Sql/servers/**/databases/master' under resource group 'rg-**' was not found. For more details please go to https://aka.ms/ARMResourceNotFoundFix"

with module.sql.module.server.azurerm_monitor_diagnostic_setting.this,
on ..\..\modules\server\main.tf line 128, in resource "azurerm_monitor_diagnostic_setting" "this":
128: resource "azurerm_monitor_diagnostic_setting" "this" {
``` 
This is the error message i get when i try to script in the "master" directly into the target resource id.

```pwsh
 Error:          Received unexpected error:
FatalError{Underlying: error while running command: exit status 1; ╷
 Error: Database "master" (Resource Group "rg-**", SQL Server "sql-**") was not found
 with module.sql.module.server.data.azurerm_mssql_database.master, on ..\..\modules\server\main.tf line 123, in data "azurerm_mssql_database" "master": 123:  data "azurerm_mssql_database" "master" {
```
And this is the original first error i got.
#### Reference
- https://github.com/hashicorp/terraform-provider-azurerm/blob/main/examples/sql-azure/sql_auditing_log_analytics/main.tf
- https://developer.hashicorp.com/terraform/language/meta-arguments/lifecycle
- https://developer.hashicorp.com/terraform/language/expressions/custom-conditions#preconditions-and-postconditions

**Comment by hknutsen** [comment]
Tests pass 🎉

**Comment by hknutsen** [comment]
Will force at db-level as well.

File: main.tf
Comment (kov117) [inline main.tf:1]: how did you figure out the solution for

**Review by kov117 — COMMENTED** [review]
looks good and terratest succesfull 👍🏽 quick question, how did you figure out the solution to the issue with master database ?

**Comment by hknutsen** [comment]
@kov117 I just thought that since configuration of server level auditing is done at the master database scope, all resources related to the audit configuration would have a dependency on said master database.

Since `azurerm_mssql_server_extended_auditing_policy` is required to configure server level auditing, it made sense to me that this resource would need to have some hidden dependency on the master database as well, and thus there would have to be some logic in the Azure provider to ensure the master database is created before this resource.

Thus, a logical solution would be to add an explicit dependency for `azurerm_monitor_diagnostic_setting`, to ensure that it will be created after `azurerm_mssql_server_extended_auditing_policy`, ensuring that the master database has been created before trying to create a diagnostic setting for it.

Basically: `azurerm_monitor_diagnostic_setting` > (explicit dependency) > `azurerm_mssql_server_extended_auditing_policy` > (hidden dependency) > master database, resulting in the diagnostic setting depending on the master database.

**Review by helenakallekleiv — APPROVED** [review]
Tests pass 👍🏼
